include/hip/devicelib/type_casting_intrinsics.hh

```diff
@@ -98,4 +98,21 @@ EXPORT float __ull2float_rn(unsigned long long int x);
 EXPORT float __ull2float_ru(unsigned long long int x);
 EXPORT float __ull2float_rz(unsigned long long int x);
 
+#if defined(__HIP_DEVICE_COMPILE__)
+EXPORT long long int __double_as_longlong(double x) {
+  static_assert(sizeof(long long) == sizeof(double), "");
+  long long Tmp;
+  __builtin_memcpy(&Tmp, &x, sizeof(Tmp));
+  return Tmp;
+}
+
+EXPORT double __longlong_as_double(long long int x) {
+  static_assert(sizeof(double) == sizeof(long long), "");
+  double Tmp;
+  __builtin_memcpy(&Tmp, &x, sizeof(Tmp));
+  return Tmp;
+}
+
+#endif
+
```

Comment:
It would make more sense for me that implementation would go into the implementation file (`devicelib.cl`) and we just leave headers here. The current state of the device library might be confusing so I could write a doc explaining it. In short, I wanted to split the device library into headers and implementations and get rid of all device library functions being implemented in macros (which was inherited from HIPLZ) and very difficult to read and/or debug individual functions. 

I emailed Rahul to get clarification as to why doesn't HIP provide these headers.